apollo-federation/src/sources/connect/json_selection/apply_to.rs

```diff
@@ -148,27 +155,60 @@ impl ApplyToError {
     }
 }
 
+// Rust doesn't allow implementing methods directly on tuples like
+// (Option<JSON>, Vec<ApplyToError>), so we define a trait to provide the
+// methods we need, and implement the trait for the tuple in question.
+pub(super) trait ApplyToResultMethods {
+    fn prepend_errors(self, errors: &[ApplyToError]) -> Self;
+
+    fn and_then_collecting_errors(
+        &self,
+        f: impl FnOnce(&JSON) -> (Option<JSON>, Vec<ApplyToError>),
+    ) -> (Option<JSON>, Vec<ApplyToError>);
+}
+
+impl ApplyToResultMethods for (Option<JSON>, Vec<ApplyToError>) {
+    // Intentionally taking ownership of self to avoid cloning, since we pretty
+    // much always use this method to replace the previous (value, errors) tuple
+    // before returning.
+    fn prepend_errors(self, errors: &[ApplyToError]) -> Self {
+        if errors.is_empty() {
+            self
+        } else {
+            let (value_opt, apply_errors) = self;
+            (value_opt, [errors, apply_errors.as_slice()].concat())
+        }
+    }
+
+    // A substitute for Option<_>::and_then that accumulates errors behind the
+    // scenes. I'm no Haskell programmer, but this feels monadic? ¯\_(ツ)_/¯
+    fn and_then_collecting_errors(
+        &self,
```

Comment:
This would also take `self`, I think, if we switch to passing owned values. Which means that `f` _could_ take an owned value if that's useful (would prevent cloning?)

Reply:
Yes that works: https://github.com/apollographql/router/commit/b04f54241bf4760b92a3c197d15b580f16ce149f